File: SpreadSimulator/Quadtree/Quadtree.c
#include "Quadtree.h"

Quadtree* createQuadTree(Rect* boundary)
{
    Quadtree* tree = calloc(1, sizeof(Quadtree));

    tree->boundary = *boundary;
    tree->ipPairsSize = 0;

    tree->northWest = NULL;
    tree->northEast = NULL;
    tree->southWest = NULL;
    tree->southEast = NULL;

    return tree;
}

bool insert(Quadtree* tree, PositionIndexPair pair)
{
    if (!containsPoint(&tree->boundary, &pair.position))
        return false;

    if (tree->ipPairsSize < NODE_CAPACITY && tree->northWest == NULL)
    {
        tree->ipPairs[tree->ipPairsSize++] = pair;
        return true;
    }

    if (tree->northWest == NULL)
    {
        subdivide(tree);
    }

    if (insert(tree->northWest, pair))
        return true;
    if (insert(tree->northEast, pair))
        return true;
    if (insert(tree->southWest, pair))
        return true;
    if (insert(tree->southEast, pair))
        return true;

    return false;
}

void subdivide(Quadtree* tree)
{
    Rect tmp_rect;

    // NORTH WEST
    Vector2 northWestPos = createVector(
            tree->boundary.center.x - tree->boundary.halfDim.x / 2,
            tree->boundary.center.y - tree->boundary.halfDim.y / 2);
    Vector2 northWestDim =
            createVector(tree->boundary.halfDim.x / 2, tree->boundary.halfDim.y / 2);

    tmp_rect = createRect(&northWestPos, &northWestDim);
    tree->northWest = createQuadTree(&tmp_rect);

    // NORTH EAST
    Vector2 northEastPos = createVector(
            tree->boundary.center.x + tree->boundary.halfDim.x / 2,
            tree->boundary.center.y - tree->boundary.halfDim.y / 2);
    Vector2 northEastDim =
            createVector(tree->boundary.halfDim.x / 2, tree->boundary.halfDim.y / 2);

    tmp_rect = createRect(&northEastPos, &northEastDim);
    tree->northEast = createQuadTree(&tmp_rect);

    // SOUTH WEST
    Vector2 southWestPos = createVector(
            tree->boundary.center.x - tree->boundary.halfDim.x / 2,
            tree->boundary.center.y + tree->boundary.halfDim.y / 2);
    Vector2 southWestDim =
            createVector(tree->boundary.halfDim.x / 2, tree->boundary.halfDim.y / 2);

    tmp_rect = createRect(&southWestPos, &southWestDim);
    tree->southWest = createQuadTree(&tmp_rect);

    // SOUTH EAST
    Vector2 southEastPos = createVector(
            tree->boundary.center.x + tree->boundary.halfDim.x / 2,
            tree->boundary.center.y + tree->boundary.halfDim.y / 2);
    Vector2 southEastDim =
            createVector(tree->boundary.halfDim.x / 2, tree->boundary.halfDim.y / 2);

    tmp_rect = createRect(&southEastPos, &southEastDim);
    tree->southEast = createQuadTree(&tmp_rect);
}

PositionIndexPairList queryRange(Quadtree* tree, Rect* range)
{
    PositionIndexPairList list = createPositionIndexPairList(4);

    if (!intersects(&tree->boundary, range))
        return list;

    for (int i = 0; i < tree->ipPairsSize; i++)
    {
        if (containsPoint(range, &tree->ipPairs[i].position))
        {
            list = pushPositionIndexPair(list, tree->ipPairs[i]);
        }
    }

    if (tree->northWest == NULL)
        return list;

    list = appendPositionIndexPairList(list, queryRange(tree->northWest, range));
    list = appendPositionIndexPairList(list, queryRange(tree->northEast, range));
    list = appendPositionIndexPairList(list, queryRange(tree->southWest, range));
    list = appendPositionIndexPairList(list, queryRange(tree->southEast, range));

    return list;
}

void freeQuadtree(Quadtree* tree)
{
    if (!tree)
        return;
    freeQuadtree(tree->northWest);
    freeQuadtree(tree->northEast);
    freeQuadtree(tree->southWest);
    freeQuadtree(tree->southEast);

    free(tree);
}

File: quadtree/Quadtree.go
package quadtree

import (
	"SpreadSimulator/util"
)

//Quadtree constants
const (
	NodeCapacity = 4
)

//Quadtree struct
type Quadtree struct {
	boundary util.Rect

	entries []util.PositionIndexPair

	northWest *Quadtree
	northEast *Quadtree
	southWest *Quadtree
	southEast *Quadtree
}

//NewQuadtree creates a new Quadtree
func NewQuadtree(boundary util.Rect) *Quadtree {
	return &Quadtree{
		boundary,
		make([]util.PositionIndexPair, 0),
		nil, nil, nil, nil,
	}
}

//Insert function returns true on success
func (qtree *Quadtree) Insert(pair util.PositionIndexPair) bool {
	if !qtree.boundary.ContainsPoint(pair.Position) {
		return false
	}

	if len(qtree.entries) < NodeCapacity && qtree.northWest == nil {
		qtree.entries = append(qtree.entries, pair)
		return true
	}

	if qtree.northWest == nil {
		qtree.subdivide()
	}

	if qtree.northWest.Insert(pair) {
		return true
	}
	if qtree.northEast.Insert(pair) {
		return true
	}
	if qtree.southWest.Insert(pair) {
		return true
	}
	if qtree.southEast.Insert(pair) {
		return true
	}

	//Should not happen
	return false
}

//QueryRange returns all indexed points within the given range
func (qtree *Quadtree) QueryRange(bounds util.Rect) []int {
	inRange := make([]int, 0)

	if !qtree.boundary.IntersectsRect(bounds) {
		return inRange
	}

	for i := range qtree.entries {
		if bounds.ContainsPoint(qtree.entries[i].Position) {
			inRange = append(inRange, qtree.entries[i].Index)
		}
	}

	if qtree.northWest == nil {
		return inRange
	}

	inRange = append(inRange, qtree.northWest.QueryRange(bounds)...)
	inRange = append(inRange, qtree.northEast.QueryRange(bounds)...)
	inRange = append(inRange, qtree.southWest.QueryRange(bounds)...)
	inRange = append(inRange, qtree.southEast.QueryRange(bounds)...)

	return inRange
}

func (qtree *Quadtree) subdivide() {
	NWCenter := util.Vector2f{
		X: qtree.boundary.Center.X - qtree.boundary.HalfDim.X/2,
		Y: qtree.boundary.Center.Y - qtree.boundary.HalfDim.Y/2,
	}
	qtree.northWest = NewQuadtree(util.NewRect(NWCenter, util.Vector2f{X: qtree.boundary.HalfDim.X / 2, Y: qtree.boundary.HalfDim.Y / 2}))

	NECenter := util.Vector2f{
		X: qtree.boundary.Center.X + qtree.boundary.HalfDim.X/2,
		Y: qtree.boundary.Center.Y - qtree.boundary.HalfDim.Y/2,
	}
	qtree.northEast = NewQuadtree(util.NewRect(NECenter, util.Vector2f{X: qtree.boundary.HalfDim.X / 2, Y: qtree.boundary.HalfDim.Y / 2}))

	SWCenter := util.Vector2f{
		X: qtree.boundary.Center.X - qtree.boundary.HalfDim.X/2,
		Y: qtree.boundary.Center.Y + qtree.boundary.HalfDim.Y/2,
	}
	qtree.southWest = NewQuadtree(util.NewRect(SWCenter, util.Vector2f{X: qtree.boundary.HalfDim.X / 2, Y: qtree.boundary.HalfDim.Y / 2}))

	SECenter := util.Vector2f{
		X: qtree.boundary.Center.X + qtree.boundary.HalfDim.X/2,
		Y: qtree.boundary.Center.Y + qtree.boundary.HalfDim.Y/2,
	}
	qtree.southEast = NewQuadtree(util.NewRect(SECenter, util.Vector2f{X: qtree.boundary.HalfDim.X / 2, Y: qtree.boundary.HalfDim.Y / 2}))
}
